topdown/trace_test.go

```diff
@@ -1242,3 +1242,48 @@ query:1      | Redo data.test = _
 	PrettyTraceWithLocation(&buf, *tracer)
 	compareBuffers(t, expected, buf.String())
 }
+
+func TestPrettyTraceWithUnifyOps(t *testing.T) {
+	module := `package test
+
+	p[x] { x = 1 }`
+
+	ctx := context.Background()
+	compiler := compileModules([]string{module})
+	store := inmem.NewFromObject(nil)
+	txn := storage.NewTransactionOrDie(ctx, store)
+	defer store.Abort(ctx, txn)
+
+	tracer := NewBufferTracer()
+	query := NewQuery(ast.MustParseBody("data.test.p")).
+		WithCompiler(compiler).
+		WithStore(store).
+		WithTransaction(txn).
+		WithTracer(tracer)
+
+	_, err := query.Run(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	expected := `query:1     Enter data.test.p
+query:1     | Eval data.test.p
+            | Unify data.test.p = _
+query:1     | Index data.test.p (matched 1 rule)
+query:3     | Enter data.test.p
+query:3     | | Eval x = 1
+            | | Unify x = 1
```

Comment:
I'm not sure I understand the reasoning for not having a location here. It is somehow related to some code location, or it wouldn't happen -- isn't it? 🤔

Reply:
Well, it depends a bit.  Some unifications have a clear location, e.g. `x = 1`.  But if we have something like `[x, y] = [1, 2]`, there are three unifications happening -- is the location of all of them the same?  Unifications can also happen as part of calling a builtin, in the arguments or in the result.  Do we then use the builtin location?

If we wanted to do that, we would need to add an extra argument to `biunify`, and its many variants like `biunifyValues`, `biunifyArraysRec`, ... and then pass it on recursively.  Since these really form the core of the interpreter I was a bit concerned about performance there, so I opted to leave it out for now.

But while writing out this comment I realized that we always want the location of the last event.  So maybe I can just store it in `traceEvent` for the next call.  I'll give that a try.

Reply:
@srenatus I've added something like that here: https://github.com/open-policy-agent/opa/pull/5203/commits/25f731d05b3848bbaeea7ce65833179a0f6ef681

The alternative is doing something like this: https://github.com/jaspervdj/opa/commit/f4bff1dc29945f0bf9e7f12d74b7fd71dc32a980, but that seems to get messy fast especially since we're saving/unsaving unifications, so the extra info needs to be saved as well.  I can try to push that further though if there's a preference for it.